web_app/utils/login.py

```diff
@@ -1,69 +1,123 @@
 """Handle user log-in and account info."""
 
 import logging
-from typing import Any, Dict, Optional
+from typing import cast, Dict, Final
 
+import ldap3  # type: ignore[import]
 from flask_login import UserMixin  # type: ignore[import]
 
-from ..config import login_manager
+from ..config import ADMINS, login_manager, user_lookup_cache
 
+_LDAP_BASE: Final[str] = "ou=People,dc=icecube,dc=wisc,dc=edu"
+_LDAP_URI: Final[str] = "ldaps://ldap-1.icecube.wisc.edu"
 
-def _mock_account_lookup(email: str) -> Dict[str, Any]:
-    name = email.split("@")[0]
-    if email.split("@")[1].upper() == "ADMIN":
-        response = {"name": name, "is_admin": True}
-    else:
-        inst = email.split("@")[1].upper()
-        response = {"institution": inst, "name": name}
 
-    return response
+def _ldap_server() -> ldap3.Server:
+    return ldap3.Server(_LDAP_URI, connect_timeout=5)
+
+
+def _ldap_uid(uid: str) -> str:
+    """Little hack to launch into non-admin mode using an admin account."""
+    return uid.strip("~")
+
+
+class InvalidLoginException(Exception):
+    """Exception for an invalid login attempt."""
+
+
+def _ldap_lookup_user(uid: str) -> "User":
+    """Look up user info via LDAP."""
+    user = User()
+    user.id = uid  # use uid as the id
+
+    conn = ldap3.Connection(_ldap_server(), auto_bind=True)
+    conn.search(_LDAP_BASE, f"(uid={_ldap_uid(user.id)})", attributes=["*"])
+    info = conn.entries[0]
+    # logging.warning(f"{info=}")
+
+    # name
+    user.name = cast(str, info.cn.value)
+    try:  # try to build full nickname
+        if info.i3NickName.value:
+            user.name = f"{info.i3NickName.value} {info.sn.value}"
+    except ldap3.core.exceptions.LDAPCursorAttributeError:
+        pass
+
+    # institution
+    # user.institution = ... # TODO: now, this is set in login() callback
+
+    # admin status
+    user.is_admin = user.id in ADMINS
+
+    return user
+
+
+def _ldap_try_login(uid: str, pwd: str) -> None:
+    """Try to get an LDAP connection.
+
+    https://github.com/WIPACrepo/iceprod/blob/master/iceprod/server/rest/auth.py#L454
+    https://www.programcreek.com/python/example/107948/ldap3.Connection
+    """
+    try:
+        _ = ldap3.Connection(
```

Comment:
is the `_ = ` for linting?  normally I'd just remove that if I didn't care about the return value.

Reply:
that's a habit I've gotten into, I don't know where I picked it up